### PR TITLE
Fix!: Improve handling of the `--where` option in table_diff

### PR DESCRIPTION
**Pull request by erindru**
Prior to this change, running `sqlmesh table_diff` with something like `--where "foo = 'bar'"` would throw an error like:

```
duckdb.duckdb.BinderException: Binder Error: Ambiguous reference to column name "foo" (use: "t.foo" or "s.foo")
```

This was because the where clause was injected verbatim without automatically qualifying the columns. The user had to know that table_diff used a subselect internally and supply `--where "s.foo = 'bar' and t.foo = 'bar'"` in order to get this to work.

This PR refactors the generated table_diff query to use CTE's to build the source / target / stats datasets so the `--where` clause can still be injected verbatim but it's now scoped to each dataset so there are no more ambiguous column references

### CODE REVIEW

File: sqlmesh/core/table_diff.py [hunk 1]
```diff
@@ -272,7 +272,22 @@ def _column_expr(name: str, table: str) -> exp.Expression:
             def name(e: exp.Expression) -> str:
                 return e.args["alias"].sql(identify=True)
 
-            query = (
+            source_query = (
+                exp.select(*(exp.column(c) for c in source_schema.keys()))
```
Comment (tobymao) [inline sqlmesh/core/table_diff.py:276]: you can just pass c, if it’s a string

you don’t need keys(), dict is iterable

File: sqlmesh/core/table_diff.py [hunk 2]
```diff
@@ -272,7 +272,22 @@ def _column_expr(name: str, table: str) -> exp.Expression:
...
+                .from_(self.source)
+                .where(self.where)
+            )
+            target_query = (
+                exp.select(*(exp.column(t) for t in target_schema.keys()))
```
Comment (tobymao) [inline sqlmesh/core/table_diff.py:281]: ditto

File: sqlmesh/core/table_diff.py [hunk 3]
```diff
@@ -313,14 +328,16 @@ def name(e: exp.Expression) -> str:
...
                 )
-                .from_(exp.alias_(self.source, "s"))
-                .join(
-                    self.target,
-                    on=self.on,
-                    join_type="FULL",
-                    join_alias="t",
-                )
-                .where(self.where)
+                .from_(exp.alias_(source_identifier, "s"))
```
Comment (tobymao) [inline sqlmesh/core/table_diff.py:331]: doesn’t seem right since it’s an identifier, is there no error here?

File: sqlmesh/core/table_diff.py [hunk 4]
```diff
@@ -313,14 +328,16 @@ def name(e: exp.Expression) -> str:
...
+                .join(exp.alias_(target_identifier, "t"), on=self.on, join_type="FULL")
+            )
+
```
Comment (tobymao) [inline sqlmesh/core/table_diff.py:334]: i don’t like manually creating with, can you use a builder api or use subqueries?

File: sqlmesh/core/table_diff.py (hunk 3 above)
Comment (erindru) [reply sqlmesh/core/table_diff.py:331]:
```
>>> exp.alias_(exp.to_identifier("__source"), "s").sql()
'__source AS s'
```

It got the correct result. What should I have done?

File: sqlmesh/core/table_diff.py (hunk 4 above)
Comment (erindru) [reply sqlmesh/core/table_diff.py:334]: Is the builder API documented anywhere? There didnt seem to be a `.with(..)`

File: sqlmesh/core/table_diff.py (hunk 3 above)
Comment (tobymao) [reply sqlmesh/core/table_diff.py:331]: exp.tables(…).as_()

File: sqlmesh/core/table_diff.py (hunk 1 above)
Comment (erindru) [reply sqlmesh/core/table_diff.py]: It looks like I still need to use `exp.column()` due to this effect:

```
>>> exp.select("country code").sql()
'SELECT country AS code'

>>> exp.select(exp.column("country code")).sql()
'SELECT "country code"'
```

however i've removed `.keys()`

File: sqlmesh/core/table_diff.py (hunk 4 above)
Comment (erindru) [reply sqlmesh/core/table_diff.py:334]: ah, I think I found it. There appears to be no top-level `with_()` like there is for `select()` but doing `exp.Select().with_(...)` gets me there.

Is that better?

Comment (tobymao) [reply sqlmesh/core/table_diff.py:334]: yea

**Review by georgesittas — APPROVED** [review]

**Pull request merged**